src/versions.cpp

```diff
@@ -0,0 +1,272 @@
+    if (pos <= start || pos - start >= 20) {
+        return std::string::npos;
+    }
+
+    return pos;
+}
+
+void find_des(const std::string& line,
+              const std::string& str,
+              const std::string& searchee,
+              std::set<std::string>& results)
+{
+    size_t j = 0;
+    while (j < str.size()) {
+        size_t i = str.find(searchee, j);
+        if (i == std::string::npos) {  // wasnt found
+            break;
+        }
+        j = i + searchee.size();
+
+        size_t end = get_suffix(str, j);
+        size_t start = get_des_prefix(str, i);
+
+        // there must be at least suffix or at least prefix
+        if (end == std::string::npos && start == std::string::npos) {
+            continue;
+        }
+
+        if (end == std::string::npos) {
+            end = j;
+        }
+
+        if (start == std::string::npos) {
+            start = i;
+        }
+
+        results.insert(line.substr(start, end - start));
+        j = end;
+    }
+}
+
+
+void find_versions(const std::string& line,
+                   const std::string& str,
+                   const std::string& searchee,
+                   std::set<std::string>& results) 
+{
+    size_t j = 0;  // one ofter the end of the last match
+    while(j < str.size()) {
+        size_t i = str.find(searchee, j); // begining of the matched string
+        if (i == std::string::npos) {  // wasnt found
+            break;
+        }
+        j = i + searchee.size();
+
+        // must be the begining of a word
+        if (i != 0 && !is_space(str[i - 1]) && str[i - 1] != '(') {
+            continue;
+        }
+
+        size_t end = get_suffix(str, j);
+        if (end == std::string::npos) {
+            continue;
+        }
+        
+        results.insert(line.substr(i, end - i));
+        j = end;
+    }
+}
+
+
+nlohmann::json parse_versions(const std::vector<std::string>& data) {
+    std::string str;
+
+    std::set<std::string> eal;
+    std::set<std::string> rsa;
+    std::set<std::string> ecc;
+    std::set<std::string> sha;
+    std::set<std::string> global_platfom;
+    std::set<std::string> java_card;
+    std::set<std::string> des;
+
+    size_t line_no = 0;
+    for (const auto& line : data) {
+        str.clear();
+        for (size_t i = 0; i < line.size(); ++i) {
+            str.push_back(to_lower(line[i]));
+        }
+
+        find_des(line, str, "des", des);
+        find_versions(line, str, "eal", eal);
+        find_versions(line, str, "sha", sha);
+        find_versions(line, str, "rsa", rsa);
+        find_versions(line, str, "ecc", rsa);
+        find_versions(line, str, "java card", java_card);
+        find_versions(line, str, "global platform", global_platfom);
+
+        line_no++;
+    }
+
+    js::json js;
+
+    if (!eal.empty()) js["eal"] = js::json(eal);
+    if (!rsa.empty()) js["rsa"] = js::json(rsa);
+    if (!sha.empty()) js["sha"] = js::json(sha);
+    if (!ecc.empty()) js["ecc"] = js::json(ecc);
```

Comment:
Probably copy & paste error but finding ecc is using rsa variable. It also results that if statement on line 262 is always false (caught by cppcheck).
```suggestion
        find_versions(line, str, "rsa", rsa);
        find_versions(line, str, "ecc", ecc);
        find_versions(line, str, "java card", java_card);
        find_versions(line, str, "global platform", global_platfom);

        line_no++;
    }

    js::json js;

    if (!eal.empty()) js["eal"] = js::json(eal);
    if (!rsa.empty()) js["rsa"] = js::json(rsa);
    if (!sha.empty()) js["sha"] = js::json(sha);
    if (!ecc.empty()) js["ecc"] = js::json(ecc);
```